src/main/java/com/aws/greengrass/deployment/ThingGroupHelper.java

```diff
@@ -102,4 +105,41 @@ public Optional<Set<String>> listThingGroupsForDevice(int maxAttemptCount) throw
             return Optional.of(thingGroupNames);
         }, "get-thing-group-hierarchy", logger);
     }
+
+    public void waitForReconnect(long timeoutMillis) throws Exception {
+        if (!deviceConfiguration.isDeviceConfiguredToTalkToCloud()) {
+            return;
+        }
+        Duration initialInterval = Duration.ofMillis(timeoutMillis / 8);
+        Duration maxRetryInterval = Duration.ofMillis(timeoutMillis / 4);
+
+        try {
+            RetryUtils.runWithRetry(clientExceptionRetryConfig.toBuilder()
+                            .maxAttempt(3)
+                            .initialRetryInterval(initialInterval)
+                            .maxRetryInterval(maxRetryInterval)
+                            .build(),
+                    () -> {
+                        ListThingGroupsForCoreDeviceRequest request = ListThingGroupsForCoreDeviceRequest.builder()
+                                .coreDeviceThingName(Coerce.toString(deviceConfiguration.getThingName()))
+                                .build();
+
+                        ListThingGroupsForCoreDeviceResponse response;
+                        try {
+                            response =
+                                    clientFactory.fetchGreengrassV2DataClient().listThingGroupsForCoreDevice(request);
+                        } catch (GreengrassV2DataException e) {
+                            if (RetryUtils.retryErrorCodes(e.statusCode())) {
+                                throw new RetryableServerErrorException("Failed with retryable error " + e.statusCode()
+                                        + " when calling listThingGroupsForCoreDevice", e);
+                            }
+                            throw e;
+                        }
+                        return response;
+                    }, "get-thing-group-hierarchy", logger);
+        } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Consider passing the exception object either as  &#34;rootCause&#34; or as an inner exception parameter, to the constructor of the new exception before throwing new exception. This will help keep stack trace.
```
@@ -141,3 +141,3 @@
             throw new ComponentConfigurationValidationException(&#34;HTTP client failed to reconnect with new configuration&#34;,
-                    DeploymentErrorCode.FAILED_TO_RECONNECT);
+                    DeploymentErrorCode.FAILED_TO_RECONNECT, e);
         }

```

Reply:
+1